File: scripts/process_imperial.py
import os
import json
import re

dresses = {
    'Alma': {'has_face': True},
    'Argente': {'has_face': True},
    'Aurelia': {'has_face': False},
    'Daiana': {'has_face': True},
    'Elia': {'has_face': True},
    'Elise': {'has_face': True},
    'Evora': {'has_face': True},
    'Ivory Grace': {'has_face': True},
    'Lumina': {'has_face': True},
    'Mayson': {'has_face': True},
    'Queen': {'has_face': True},
    'Selena': {'has_face': True},
}

with open('data/higgsfield-elements.json', 'r') as f:
    data = json.load(f)

for item in data:
    if item['collection'] == 'Imperial':
        name = item['dress']
        if name in dresses:
            has_face = dresses[name]['has_face']
            item['has_face_in_primary'] = has_face
            item['cleaned'] = True
            
            if has_face:
                # crop the image
                import urllib.parse
                img_url = item['images'][0]
                img_path = img_url.replace('https://salonfya.com', '')
                local_in = os.path.join('public', urllib.parse.unquote(img_path.lstrip('/')))
                
                # generate out path
                base_name = os.path.basename(local_in)
                name_no_ext, ext = os.path.splitext(base_name)
                out_name = f"{name_no_ext}_nohead{ext}"
                local_out = os.path.join('public', 'images', 'clean', out_name)
                
                # call crop_head.py
                os.system(f"python3 scripts/crop_head.py \"{local_in}\" \"{local_out}\"")
                
                # update the catalog.ts with the new image
                # this is tricky with python, better to write a separate node script for catalog update
                dresses[name]['new_img'] = f"/images/clean/{out_name}"

with open('data/higgsfield-elements.json', 'w') as f:
    json.dump(data, f, indent=2)

# Write a node script to update catalog.ts
node_script = """
import * as fs from 'fs';
import * as path from 'path';

const catalogPath = path.join(process.cwd(), 'src/data/catalog.ts');
let content = fs.readFileSync(catalogPath, 'utf8');

const updates = """ + json.dumps(dresses) + """;

const col = 'IMPERIAL';
const regex = new RegExp(`const ${col}_SEEDS: Seed\\[\\] = \\\\s*\\\\[([\\\\s\\\\S]*?)\\\\];`, 'g');

content = content.replace(regex, (match, arrayContent) => {
  let newArrayContent = arrayContent;
  
  for (const name in updates) {
    if (updates[name].new_img) {
      const newImg = updates[name].new_img;
      // find the dress block
      // regex to find `name: 'name', ... images: [ ... ]`
      const dressRegex = new RegExp(`(name:\\s*'${name}',[\\\\s\\\\S]*?images:\\s*\\[\\s*)([^,]+)(,[\\\\s\\\\S]*?\\])`, 'g');
      newArrayContent = newArrayContent.replace(dressRegex, (m, p1, p2, p3) => {
        // p2 is the first image, e.g. flat('alma_front.jpg')
        // we replace it with `p('${newImg}')` wait no, `flat` or `p` are helpers.
        // It's better to just put `flat('clean/${out_name}')` if we saved to clean/
        const baseName = newImg.split('/').pop();
        return `${p1}p('clean', '${baseName}')${p3}`;
      });
    }
  }
  return `const ${col}_SEEDS: Seed[] = [${newArrayContent}];`;
});

fs.writeFileSync(catalogPath, content, 'utf8');
console.log("Catalog updated!");
"""

with open('scripts/update_imperial_catalog.ts', 'w') as f:
    f.write(node_script)

File: scripts/process_imperial.js
import fs from 'fs';
import path from 'path';
import sharp from 'sharp';

const IMPERIAL_DRESSES = [
  {
    name: 'Alma',
    images: ['alma_front.jpg', 'alma_back.jpg', 'alma_detail.jpg']
  },
  {
    name: 'Argente',
    images: ['argente_front.jpg', 'argente_front_2.jpg', 'argente_back.jpg', 'argente_detail.jpg']
  },
  {
    name: 'Aurelia',
    images: ['aurelia_front.jpg', 'aurelia_back.jpg', 'aurelia_detail.jpg', 'aurelia_closeup.jpg']
  },
  {
    name: 'Daiana',
    images: ['daiana_front.jpg', 'daiana_side.jpg', 'daiana_back.jpg', 'daiana_detail_1.jpg', 'daiana_detail_2.jpg']
  },
  {
    name: 'Elia',
    images: ['IMPERIAL/Elia/IMG_5552.jpg', 'IMPERIAL/Elia/IMG_5553.jpg', 'IMPERIAL/Elia/IMG_5554.jpg', 'IMPERIAL/Elia/IMG_5414.jpg', 'IMPERIAL/Elia/IMG_5415.jpg']
  },
  {
    name: 'Elise',
    images: ['elise_front.jpg', 'elise_side.jpg', 'elise_back.jpg', 'elise_detail.jpg']
  },
  {
    name: 'Evora',
    images: ['IMPERIAL/Evora/IMG_5580.jpg', 'IMPERIAL/Evora/IMG_5581.jpg', 'IMPERIAL/Evora/IMG_5583.jpg', 'IMPERIAL/Evora/IMG_5587.jpg']
  },
  {
    name: 'Ivory Grace',
    images: ['IMPERIAL/Ivory Grace/IMG_5395.jpg', 'IMPERIAL/Ivory Grace/IMG_5399.jpg', 'IMPERIAL/Ivory Grace/f202fbc6-6dd7-442b-8121-0cc159e5ff40.JPG']
  },
  {
    name: 'Lumiere',
    images: ['IMPERIAL/Lumiere/IMG_5439.jpg']
  },
  {
    name: 'Mayson',
    images: ['IMPERIAL/Mayson/IMG_5364.jpg', 'IMPERIAL/Mayson/IMG_5366.jpg', 'IMPERIAL/Mayson/IMG_5369.jpg', 'IMPERIAL/Mayson/IMG_5370.jpg']
  },
  {
    name: 'Queen',
    images: ['IMPERIAL/Queen/IMG_5568.jpg', 'IMPERIAL/Queen/IMG_5571.jpg', 'IMPERIAL/Queen/IMG_5463.jpg']
  },
  {
    name: 'Serena',
    images: ['IMPERIAL/Serena/IMG_5423.jpg', 'IMPERIAL/Serena/IMG_5424.jpg', 'IMPERIAL/Serena/IMG_5573.jpg', 'IMPERIAL/Serena/IMG_5460.jpg']
  }
];

const ROOT_DIR = path.resolve(process.cwd(), 'public/images');
const CLEAN_DIR = path.resolve(process.cwd(), 'public/images/clean/IMPERIAL');

async function processImages() {
  if (!fs.existsSync(CLEAN_DIR)) {
    fs.mkdirSync(CLEAN_DIR, { recursive: true });
  }

  const elements = [];

  for (const dress of IMPERIAL_DRESSES) {
    const elementName = `FYA-DRESS-IMP-${dress.name.replace(/\s+/g, '')}`;
    const cleanImagePaths = [];

    for (const imageRelPath of dress.images) {
      const sourcePath = path.join(ROOT_DIR, imageRelPath);
      const filename = path.basename(imageRelPath);
      const targetPath = path.join(CLEAN_DIR, filename);

      if (!fs.existsSync(sourcePath)) {
        console.warn(`File not found: ${sourcePath}`);
        continue;
      }

      try {
        const metadata = await sharp(sourcePath).metadata();
        const cropTop = Math.floor(metadata.height * 0.20);
        const cropHeight = metadata.height - cropTop;

        await sharp(sourcePath)
          .extract({ left: 0, top: cropTop, width: metadata.width, height: cropHeight })
          .toFile(targetPath);
        
        console.log(`Cropped: ${filename}`);
        cleanImagePaths.push(`public/images/clean/IMPERIAL/${filename}`);
      } catch (e) {
        console.error(`Error processing ${sourcePath}:`, e);
      }
    }

    elements.push({
      element_name: elementName,
      collection: 'IMPERIAL',
      images: cleanImagePaths
    });
  }

  // Update elements JSON
  const elementsJsonPath = path.resolve(process.cwd(), 'data/higgsfield-elements.json');
  let existingElements = [];
  if (fs.existsSync(elementsJsonPath)) {
    existingElements = JSON.parse(fs.readFileSync(elementsJsonPath, 'utf8'));
  }

  // Filter out any existing IMPERIAL elements to avoid duplicates during re-runs
  existingElements = existingElements.filter(e => e.collection !== 'IMPERIAL');
  const allElements = [...existingElements, ...elements];
  
  if (!fs.existsSync(path.dirname(elementsJsonPath))) {
    fs.mkdirSync(path.dirname(elementsJsonPath), { recursive: true });
  }

  fs.writeFileSync(elementsJsonPath, JSON.stringify(allElements, null, 2));
  console.log('Updated higgsfield-elements.json');
}

processImages().catch(console.error);
